### PR TITLE
REF: Add Wallet screen now uses useReducer

### PR DESCRIPTION
**Pull request by marcosrdz**

### CODE REVIEW

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/h5yptvgartweldu363xojjmp2q?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/11923)

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/c3q7eaix2znuhjhpy5k2l6iil4?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/11934)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/w7nmjw3h2e5aenhue2xfle44yy?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/11936)

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/nsagd7po5n55zd7xi7nbnw2v4u?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/12194)

**Review by Overtorment — CHANGES_REQUESTED** [review]
ELI5 wtf is useReducer how to use it

**Comment by marcosrdz** [comment]
> ELI5 wtf is useReducer how to use it
> 
> 
> 
> 

Its a common pattern to make state logic centralized and predictable. U pass a string and a payload . The reducer will determine what to do outside the component and return a new state

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/tf3uaycw4bxocpqb4b3y6naw3m?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/12207)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/f72ajygou56ndd5kv5jnqbd7ry?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/12219)

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/bvgfhu4ti35xj2uopadia7f5cm?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/12220)

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

File: screen/wallets/add.tsx [hunk 1]
```diff
@@ -38,27 +45,83 @@ const BlueApp = require('../../BlueApp');
...
+  selectedIndex: number;
+  label: string;
+  selectedWalletType: ButtonSelected;
+  backdoorPressed: number;
+  entropy: string | any[] | undefined;
+  entropyButtonText: string;
+}
+
+interface Action {
+  type: string;
```
Comment (Overtorment) [inline screen/wallets/add.tsx:69]: you can define a separate type, like

```ts
type TYPES =  'SET_LOADING' | 'SET_WALLET_BASE_URI' |'SET_SELECTED_INDEX'
```

and use it instead of `string` here, this will make it impossible to make a typo when dispatching

**Comment by Overtorment** [comment]
@marcosrdz pls tell me more about it.  will it make code more maintainable in the future, or more safe? or what?

@limpbrains what do you think about this approach?

**Comment by marcosrdz** [comment]
> @marcosrdz pls tell me more about it.  will it make code more maintainable in the future, or more safe? or what?
> 
> 
> 
> @limpbrains what do you think about this approach?
> 
> 
> 
> 

Far more readable and maintainable. The multisig creation screen will be perfect to migrate to usereducer to separate business logic from UI as its currently all mixed in together

File: screen/wallets/add.tsx (hunk 1 above)
Comment (marcosrdz) [reply screen/wallets/add.tsx]: done

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/c4negcnqemyl5l3i5fvccyfqfa?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/12237)

**Comment by limpbrains** [comment]
I love useReducer, definetely better than useState in the complex screens.
Entropy screen uses it.
One cool thing - you can easily write tests for it.

**Comment by limpbrains** [comment]
But I don't think we should refactor all screens to use it. Like here there are complex interactions, parts of the state are not depends on each other. What is the issue this PR is solving?

**Comment by marcosrdz** [comment]
> But I don't think we should refactor all screens to use it. Like here there are complex interactions, parts of the state are not depends on each other. What is the issue this PR is solving?

Its a screen that i figured would be best to showcase asyncpromise and usereducer.

**Comment by marcosrdz** [comment]
Its the screen that can be used as reference for future refactoring of other screens and as a "guide" for new future screens

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/qd3mrcme6o4uaqf2kpggtrljka?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/12245)

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @limpbrains] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/limpbrains

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Review by limpbrains — APPROVED** [review]

**Comment by GladosBlueWallet** [comment]
Wake the fuck up samurai, we have PRs to merge

image

[all PRs for @Overtorment] https://github.com/BlueWallet/BlueWallet/pulls/review-requested/Overtorment

**Review by Overtorment — APPROVED** [review]

**Comment by Overtorment** [comment]
Let's merge it, but not rush to rewrite everything with this approach

**Comment by GladosBlueWallet** [comment]
Unbelievable. You, [subject name here], must be the pride of [subject hometown here]!
